### PR TITLE
issue #93 proposition

### PR DESCRIPTION
**Pull request by cmarchand**
I've changed file generation to put them in the same directory of the compiled XSL.
Nothing tested for XQuery. Bug do not concern XQuery XSpec

### CODE REVIEW

File: src/compiler/generate-xspec-tests.xsl [hunk 1]
```diff
@@ -57,6 +57,8 @@
     <namespace-alias stylesheet-prefix="__x" result-prefix="xsl" />
     <variable name="x:stylesheet-uri" as="xs:string" select="'{$stylesheet-uri}'" />
   	<output name="x:report" method="xml" indent="yes" />
+    <!-- issue #93 -->
+    <variable name="x:compiled-xsl-uri" select="static-base-uri()" as="xs:anyURI"/>
```
Comment (AirQuick) [inline src/compiler/generate-xspec-tests.xsl]: Assuming the variable is private, isn't the namespace prefix `impl` more suitable than `x`?

**Comment by AirQuick** [comment]
`test/xspec-50.xspec` and `test/xspec-55.xspec` perform [test on `test:report-value()`](https://travis-ci.org/xspec/xspec/jobs/208241260). You need to update them accordingly or make the parameter optional.

File: src/compiler/generate-xspec-tests.xsl (hunk 1 above)
Comment (cmarchand) [inline src/compiler/generate-xspec-tests.xsl]: You are correct.

**Comment by cmarchand** [comment]
After making the parameter optinal, xspec-50 & 55 still fail.
That's strange : xspec-50 & xspec-55 are on value being a simple type (xs:untypedAtomic, xs:hexBinary, xs:deciml, xs:double, xs:integer), and my change concerns only value being a node()+.

I'm gonna have a look at this.

**Comment by AirQuick** [comment]
As far as I can see,

* (a) The template in question, `test:report-value`, is designed to be called when generating the result XML. More specifically, the template is called while the transformation is running with the following parameters:
  * Source document = `*.xspec` file
  * XSLT = Compiled `*.xsl`
  * Initial template = `{http://www.jenitennison.com/xslt/xspec}main`
  * Output document = Result XML (typically `*-result.xml`)

* (b) Per [the spec](http://www.w3.org/TR/xslt20/#element-result-document) of `xsl:result-document`, "The href attribute... If the effective value is a relative URI, then it is resolved relative to the base output URI."

Considering (a) and (b), the current implementation (before merging the pull request) seems to be working correctly on the shell/batch script and on Ant. That is, the location of the `*.xml` file (for 1000+ nodes) follows the location of the result XML file (`RESULT` on the shell/batch script or `xspec.result.xml` on Ant build).

Does the Maven plugin possibly set an unusual base output URI while processing (a)?

**Comment by AirQuick** [comment]
Also, the #93 says,

> If `input.xml` and `output.xml` are 1000+ elements, temp files are generated in **current directory**.

and the Saxon implementation [doc](http://www.saxonica.com/documentation/index.html#!xsl-elements/result-document) says, (bold mine)

> xsl:result-document
> ...
> href?
> ... 
> If the base output URI is not known, then **the current directory** is used

so they seem to suggest that the transformer in the Maven plugin might be failing to establish the base outout URI when creating the result XML. What do you think?

**Comment by AirQuick** [comment]
I took `20170304_xspec-issue3.zip` from [the original issue](https://github.com/xspec/xspec-maven-plugin-1/issues/3#issuecomment-283906327) and modified `identity.xsl` as follows:

```xml
...
  <xsl:template match="document-node() | node() | @*">

	<xsl:if test="self::document-node()"
		xmlns:saxon="http://saxon.sf.net/"
		xmlns:jt="http://saxon.sf.net/java-type"
		xmlns:Controller="java:net.sf.saxon.Controller">
		<xsl:variable name="controller" select="saxon:getController()" as="jt:net.sf.saxon.Controller"/>
		<xsl:message select="'Base output URI =',Controller:getBaseOutputURI($controller)" />
	</xsl:if>
  	
  	<xsl:copy>
...
```

When I ran `20170304_xspec-issue3/src/test/xspec/identity.xspec` with the standard XSpec [command](https://github.com/xspec/xspec/tree/f4811a5976c637210e651f5b46d962f38d183f9b/bin), I got this:

> ...
> Running Tests...
> Testing with SAXON EE 9.7.0.15
> Scenario comparing to a file
> Base output URI = file:/my/path/to/20170304_xspec-issue3/src/test/xspec/xspec/identity-result.xml
> ...

and the external XML files (such as `d6.xml` and `d8.xml`) were created in the same directory as `identity-result.xml`. This is the correct behavior, I believe.

On the other hand, when I ran it with `mvn test` (with rewriting `pom.xml` to depend on Saxon-EE), I got this:

> ...
> [INFO] --- xspec-maven-plugin:1.5.0-RC2:run-xspec (default) @ issue3 ---
> ...
> [INFO] Executing XSpec: identity.xspec.xslt
> Testing with SAXON EE 9.7.0.15
> Scenario comparing to a file
> Base output URI =
> ...

Notice the empty `Base output URI =` line.

So it seems highly likely that the Maven plugin is not providing the base output URI.

**Comment by cmarchand** [comment]
You are correct, setting base output URI solve the problem.
Issue can be closed, it is not an issue, and pull request can be cancelled.
I re-open https://github.com/xspec/xspec-maven-plugin-1/issues/3

Best,
Christophe

**Comment by AirQuick** [comment]
Great.

If in the future we want to have a finer control over the location of the external XML files, this pull request will probably help. In such a case, the parameter would be provided via a global transformation parameter.

**Pull request closed without merging**